GVFS/GVFS.Platform.Windows/DiskLayoutUpgrades/DiskLayout12to13Upgrade_FolderPlaceholder.cs

```diff
@@ -81,30 +81,32 @@ private static IEnumerable<string> GetFolderPlaceholdersFromDisk(ITracer tracer,
                 {
                     if (!directory.EndsWith(Path.DirectorySeparatorChar + GVFSConstants.DotGit.Root))
                     {
-                        OnDiskFileState fileState = OnDiskFileState.Full;
-                        HResult result = Utils.GetOnDiskFileState(directory, out fileState);
-                        if (result == HResult.Ok)
+                        if (Directory.Exists(directory))
                         {
-                            if (IsPlaceholder(fileState))
+                            OnDiskFileState fileState = OnDiskFileState.Full;
+                            bool result = Utils.TryGetOnDiskFileState(directory, out fileState);
+                            if (result == true)
                             {
-                                yield return directory;
-                            }
+                                if (IsPlaceholder(fileState))
+                                {
+                                    yield return directory;
+                                }
 
-                            // Recurse into placeholders and full folders skipping the tombstones
-                            if (!IsTombstone(fileState))
-                            {
-                                foreach (string placeholderPath in GetFolderPlaceholdersFromDisk(tracer, fileSystem, directory))
+                                // Recurse into placeholders and full folders skipping the tombstones
+                                if (!IsTombstone(fileState))
                                 {
-                                    yield return placeholderPath;
+                                    foreach (string placeholderPath in GetFolderPlaceholdersFromDisk(tracer, fileSystem, directory))
+                                    {
+                                        yield return placeholderPath;
+                                    }
                                 }
                             }
-                        }
-                        else if (result != HResult.FileNotFound)
-                        {
-                            // FileNotFound is returned for tombstones when the filter is attached to the volume so we want to
-                            // just skip those folders.  Any other HResults may cause valid folder placeholders not to be written
-                            // to the placeholder database so we want to error out on those.
-                            throw new InvalidDataException($"Error getting on disk file state. HResult = {result} for {directory}");
+                            else
+                            {
+                                // May cause valid folder placeholders not to be written

```

Comment:
Basically we now throw exceptions where previously "HResult.FileNotFound" would not have.
@wilbaker can you confirm the change in the API and why this was the right move.

Reply:
@kewillford it looks like you added this check in PR 336155 (in Azure Repos).  Do you remember how we were getting in a state where `fileSystem.EnumerateDirectories(path)` was returning a tombstone?

@cgallred what should `Utils.TryGetOnDiskFileState` return for a path that's a tombstone when the filter is attached?

Reply:
I don't remember exactly but in reading through this maybe the previous comment should have been "when the filter is **not** attached" because you are right that the enumeration should not even return the directories that are tombstones but without the filter attached it might have been since this is potentially running before the filter has been attached to the volume.

Reply:
Found the email that addressed this and it was a bug in prjflt that was fixed and this was the quick fix for the bug so I think we should be good here.  See Bug 17097791 in the Windows Azure workitems.

Reply:
@kewillford thanks for the details!

> See Bug 17097791 in the Windows Azure workitems.

It looks like 17097791 was fixed in RS5.  

@jrbriggs @jeschu1, I suspect that if we don't do something special here then a user could have an issue if:

- They are still running on Windows RS4
- They have an old repo that has not been upgraded since this upgrade step was added.

My guess it's unlikely anyone will hit this, but I still wanted to mention it so that we can decide on the risk vs maintenance cost.